server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -197,13 +202,58 @@ public final void run() {
                 }
             }
             for (int index = 0; index < shardsIts.size(); index++) {
-                final SearchShardIterator shardRoutings = shardsIts.get(index);
-                assert shardRoutings.skip() == false;
-                performPhaseOnShard(index, shardRoutings, shardRoutings.nextOrNull());
+                final SearchShardIterator searchShardIt = shardsIts.get(index);
+                assert searchShardIt.skip() == false;
+                ShardRouting shard = searchShardIt.nextOrNull();
+                TimeValue unavailableShardsTimeout = request.waitForUnassignedPrimaryShardsAllocationTimeout();
+                if (shard == null && unavailableShardsTimeout.equals(TimeValue.ZERO) == false) {
+                    delayPhaseOnShardUntilPrimaryIsStarted(index, searchShardIt);
+                } else {
+                    performPhaseOnShard(index, searchShardIt, shard);
+                }
             }
         }
     }
 
+    private void delayPhaseOnShardUntilPrimaryIsStarted(int index, SearchShardIterator searchShardIt) {
+        final int shardIndex = index;
+        startedPrimaryShardObserver.waitUntilPrimaryShardIsStarted(searchShardIt.shardId(),
+            request.waitForUnassignedPrimaryShardsAllocationTimeout(),
+            new ActionListener<>() {
+                @Override
+                public void onResponse(ShardRouting primaryShard) {
+                    fork(() -> {
+                        SearchShardIterator updatedSearchIter = createUpdatedShardIterator(searchShardIt, primaryShard);
+                        synchronized (shardItsMutex) {
+                            List<SearchShardIterator> updatedIters = new ArrayList<>(shardsIts.size());
+                            for (int i = 0; i < shardsIts.size(); i++) {
+                                updatedIters.add(shardIndex == i ? updatedSearchIter : shardsIts.get(i));
+                            }
+                            shardsIts = new GroupShardsIterator<>(updatedIters);
+                            performPhaseOnShard(shardIndex, updatedSearchIter, updatedSearchIter.nextOrNull());
+                        }
+                    });
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    fork(() -> performPhaseOnShard(shardIndex, searchShardIt, null));
+                }
+            });
+    }
+
+    private SearchShardIterator createUpdatedShardIterator(SearchShardIterator searchShardIterator,
+                                                           ShardRouting primaryShard) {
+        assert primaryShard != null && primaryShard.started();
+
+        ShardId shardId = searchShardIterator.shardId();
+
+        return new SearchShardIterator(searchShardIterator.getClusterAlias(),
```

Comment:
Can you add a comment that it works because we only add the primary shard to the iterator (this would break `expectedOps` if we add replicas too) ?